### PR TITLE
feat(js/ai,plugins/vertexai): add rerankers

### PR DESCRIPTION
**Pull request by cabljac**
This PR adds rerankers as an action. It also adds an example reranker to the vertexai plugin.

resolves https://github.com/firebase/genkit/issues/771 

Checklist (if applicable):
- [x] Tested - only manually tested so far (manually, unit tested, etc.)
- [ ] Changelog updated
- [ ] Docs updated

I will add tests and docs once this design is confirmed and approved.

### CODE REVIEW

File: js/ai/src/reranker.ts [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      content: [{ text }],
+      metadata,
+    });
+  }
+
+  /**
+   * Concatenates all `text` parts present in the document with no delimiter.
+   * @returns A string of all concatenated text parts.
+   */
+  text(): string {
```
Comment (cabljac) [inline js/ai/src/reranker.ts:68]: Document already implements this - remove it

Comment (cabljac) [reply js/ai/src/reranker.ts:68]: In fact i've reimplemented a bunch of things here that i dont have to. I will remove.

File: js/testapps/vertexai-reranker/.env.example [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# .env.example
+
+# Firebase project configuration
+PROJECT_ID=your_project_id_here
+LOCATION=your_location_here
```
Comment (cabljac) [inline js/testapps/vertexai-reranker/.env.example:5]: remove everything except project ID and location. perhaps add GOOGLE_APPLICATION_CREDENTIALS

File: js/testapps/vertexai-reranker/README.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
+# Sample Vertex AI Plugin Retriever and Indexer with Local File
```
Comment (cabljac) [inline js/testapps/vertexai-reranker/README.md:1]: TODO: update this for this testapp

File: js/testapps/vertexai-reranker/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "@genkit-ai/ai": "workspace:*",
+    "@genkit-ai/core": "workspace:*",
+    "@genkit-ai/dev-local-vectorstore": "workspace:*",
+    "@genkit-ai/dotprompt": "workspace:*",
+    "@genkit-ai/evaluator": "workspace:*",
+    "@genkit-ai/firebase": "workspace:*",
+    "@genkit-ai/flow": "workspace:*",
+    "@genkit-ai/googleai": "workspace:*",
+    "@genkit-ai/vertexai": "workspace:*",
+    "@google-cloud/bigquery": "^7.8.0",
```
Comment (cabljac) [inline js/testapps/vertexai-reranker/package.json:26]: TODO: remove deps that aren't needed here, e.g BQ

File: js/plugins/vertexai/src/reranker.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  for (const rerankOption of rerankOptions) {
+    const reranker = defineReranker(
+      {
+        name: `vertexai/${rerankOption.name || rerankOption.model}`,
+        configSchema: VertexAIRerankerOptionsSchema.optional(),
+      },
+      async (query, documents, _options) => {
+        const auth = new GoogleAuth();
```
Comment (pavelgj) [inline js/plugins/vertexai/src/reranker.ts:105]: Can these 3 lines be moved outside of defineReranker?

**Review by pavelgj — APPROVED** [review]

File: js/plugins/vertexai/src/reranker.ts (hunk 5 above)
Comment (cabljac) [reply js/plugins/vertexai/src/reranker.ts:105]: yup!

**Review by pr-Mais — APPROVED** [review]

**Pull request merged**